cocos/base/uthash.h

```diff
@@ -857,13 +857,11 @@ do {
             (HASH_BLOOM_BYTELEN)))
 
 #ifdef NO_DECLTYPE
-#define HASH_ITER(hh,head,el,tmp)                                                \
-for((el)=(head), (*(char**)(&(tmp)))=(char*)((head)?(head)->hh.next:NULL);       \
-  el; (el)=(tmp),(*(char**)(&(tmp)))=(char*)((tmp)?(tmp)->hh.next:NULL)) 
+#define HASH_ITER(hh,head,el)                                                    \
+  for( (el)=(head);  el;  (*(char**)(&(el)))=(char*)((el)?(el)->hh.next:NULL) )
 #else
-#define HASH_ITER(hh,head,el,tmp)                                                \
-for((el)=(head),(tmp)=DECLTYPE(el)((head)?(head)->hh.next:NULL);                 \
-  el; (el)=(tmp),(tmp)=DECLTYPE(el)((tmp)?(tmp)->hh.next:NULL))
+#define HASH_ITER(hh,head,el)                                                    \
+  for( (el)=(head);  el;  (el)=DECLTYPE(el)((el)?(el)->hh.next:NULL) )
```

Comment:
[line #861](https://github.com/cocos2d/cocos2d-x/pull/19760/files#diff-426cb1038ce3f085862ce780d57b0c74R861) and  [line #864](https://github.com/cocos2d/cocos2d-x/pull/19760/files#diff-426cb1038ce3f085862ce780d57b0c74R864) are where the actual changes occurred.

The rest removed only the `tmp variables`.

Reply:
I think it is needed. For example
```c++
HASH_ITER(hh, collision_handler_hash, current)
{
    ...
    HASH_DEL(collision_handler_hash,current);
}
```

As you see, current may be deleted in the loop, so use a tmp variable to record the next element to avoid crash of this situation.

Reply:
thanks to consider this!

I see...
agree with your example.
but.. I think `tmp` is unnecessary to be visible at outside scope.
so, what about this?

```suggestion
if( void* tmp = NULL );                                                          \
else                                                                             \
for((el)=(head),(tmp)=(head)?(head)->hh.next:NULL;                               \
el; (el)=DECLTYPE(el)(tmp),(tmp)=(tmp)?(tmp)->hh.next:NULL)
```

Reply:
thanks to consider this!

I see...
agree with your example.
but.. I think `tmp` is unnecessary to be visible at outside scope.
so, what about this?

```suggestion
if( void* tmp = NULL );                                                          \
else                                                                             \
for((el)=(head),(tmp)=(head)?(head)->hh.next:NULL;                               \
el; (el)=DECLTYPE(el)(tmp),(tmp)=(tmp)?(DECLTYPE(el)(tmp))->hh.next:NULL)
```

Reply:
Sorry, i don't quite understand the codes:
```

if( void* tmp = NULL );                                                          \
| else


```

Reply:
It is equivalent to
```cpp
{ // start of local scope
    void* tmp = NULL
    if (tmp);
    else
        // for(...) will here, so tmp is visible
} // end of local scope
```

Reply:
I agree to hide tmp internally, but could you please provide a full implementation. What's the `HASH_ITER ` look like?

Reply:
ok.
to see well, I show you two version (before / after).

### before
```cpp
#ifdef NO_DECLTYPE
#define HASH_ITER(hh,head,el,tmp)                                                \
for((el)=(head), (*(char**)(&(tmp)))=(char*)((head)?(head)->hh.next:NULL);       \
  el; (el)=(tmp),(*(char**)(&(tmp)))=(char*)((tmp)?(tmp)->hh.next:NULL)) 
#else
#define HASH_ITER(hh,head,el,tmp)                                                \
for((el)=(head),(tmp)=DECLTYPE(el)((head)?(head)->hh.next:NULL);                 \
  el; (el)=(tmp),(tmp)=DECLTYPE(el)((tmp)?(tmp)->hh.next:NULL))
#endif
```

### after
```cpp
#ifdef NO_DECLTYPE
#define HASH_ITER(hh,head,el)                                                    \
if(auto tmp = el);                                                               \
else                                                                             \
  for( (el)=(head), (*(char**)(&(tmp)))=(char*)((head)?(head)->hh.next:NULL);    \
    el; (el)=(tmp), (*(char**)(&(tmp)))=(char*)((tmp )?(tmp )->hh.next:NULL) )
#else
#define HASH_ITER(hh,head,el)                                                    \
if(auto tmp = el);                                                               \
else                                                                             \
  for( (el)=(head), tmp=DECLTYPE(el)((head)?(head)->hh.next:NULL);               \
    el; (el)=(tmp), tmp=DECLTYPE(el)((tmp )?(tmp )->hh.next:NULL) )
#endif
```

it is not easy to make typecast `void*` to `decltype(el)` when `NO_DECLTYPE` condition.....

so I use auto.

can I use auto when `NO_DECLTYPE` cond?

Reply:
> can I use auto when NO_DECLTYPE cond?

I am not sure. It seems the codes are designed to run in C and C++.

Reply:
by the way,
`if(auto tmp = el)` can be true. so `else` may not be executed.
it is exactly bad code....

Reply:
it is new version.

```cpp
#ifdef NO_DECLTYPE
#define HASH_ITER(hh,head,el)                                                    \
if(void* tmp = NULL);                                                            \
else                                                                             \
for((el)=(head), (*(char**)(&(tmp)))=(char*)((head)?(head)->hh.next:NULL);       \
  el; (*(char**)(&(el)))=(char*)(tmp),tmp=(el)?(el)->hh.next:NULL)
#else
#define HASH_ITER(hh,head,el)                                                    \
if(void* tmp = NULL);                                                            \
else                                                                             \
for((el)=(head), (tmp)=DECLTYPE(el)((head)?(head)->hh.next:NULL);                \
  el; (el)=DECLTYPE(el)(tmp),tmp=(el)?(el)->hh.next:NULL)
#endif
```

use the `tmp` as storage only.
and comma operator guarantees left to right evaluation.

Reply:
Why use `if`, can just define it like this:
```c++
#ifdef NO_DECLTYPE
#define HASH_ITER(hh,head,el)                                                    \
void* tmp = NULL;                                                            \
for((el)=(head), (*(char**)(&(tmp)))=(char*)((head)?(head)->hh.next:NULL);       \
  el; (*(char**)(&(el)))=(char*)(tmp),tmp=(el)?(el)->hh.next:NULL)
#else
#define HASH_ITER(hh,head,el)                                                    \
void* tmp = NULL;                                                            \                                                                           
for((el)=(head), (tmp)=DECLTYPE(el)((head)?(head)->hh.next:NULL);                \
  el; (el)=DECLTYPE(el)(tmp),tmp=(el)?(el)->hh.next:NULL)
#endif
```

??

Reply:
It cause that the `tmp` variable will be visible at outer scope.


``` cpp
/*your version*/

{ // outer scope
    HASH_ITER(...)
    // we can't declare tmp here. (redundant declaration)
}
```

but if  use `if`,

```cpp
/*if with declare*/

{ // outer scope
    HASH_ITER(...)
    // tmp not be visible, so we can declare here.
}
```

because `if with declare` means [this](https://github.com/cocos2d/cocos2d-x/pull/19760#discussion_r288414781)

Reply:
The usage is tricky, so please add some comments in the codes.

Reply:
I agree. 
However, before committing the code, I thought of it.

Rather than using an `if(void* tmp = NULL); else { }`,
I would like to ask about using an `nested-for statement like below code` that is complex but does not confuse contributors.




```cpp
// nested-for statement version
// --------------------



#ifdef NO_DECLTYPE /** @see #else section which is same code but more modern */
#define HASH_ITER(hh,head,el)                                                    \
for(bool __tmp_loop_once__=true; __tmp_loop_once__; )                            \
  for(void* __tmp_next_el__ = NULL; __tmp_loop_once__; __tmp_loop_once__=false)  \
    for((el)=(head),                                                             \
        (*(char**)(&(__tmp_next_el__)))=(char*)((head)?(head)->hh.next:NULL);    \
        el;                                                                      \
        (*(char**)(&(el)))=(char*)(__tmp_next_el__),                             \
        __tmp_next_el__=(el)?(el)->hh.next:NULL)
#else
#define HASH_ITER(hh,head,el)                                                    \
for(bool __tmp_loop_once__=true; __tmp_loop_once__; )                            \
  for(void* __tmp_next_el__ = NULL; __tmp_loop_once__; __tmp_loop_once__=false)  \
    /** @note :                                                                  \
          double nested-for loop used to declare '__tmp_next_el__'               \
          in the local scope. */                                                 \
                                                                                 \
    for( (el)=(head),                                                            \
         (__tmp_next_el__)=DECLTYPE(el)((head)?(head)->hh.next:NULL);            \
                                                                                 \
         el;                                                                     \
                                                                                 \
         (el)=DECLTYPE(el)(__tmp_next_el__),                                     \
         __tmp_next_el__=(el)?(el)->hh.next:NULL )
#endif
```

Reply:
It is hard to understand too.

Reply:
Then, should we just close this pr..?
Actually, there doesn't seem to be a big issue.
I just want to give a little help to a grateful engine.

I confuse how to contribute well.

Reply:
I think we can close this PR though you did a great thing. You can try to fix issues in the issue system, or help to review codes, or join the forum to help to other developers.

Thanks again.

Reply:
Thanks for your kindness!